### ISSUE TITLE
Non-exported Props interface/type shouldn't be flagged by no-unused-vars

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [ota-meshi/eslint-plugin-astro#194]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.

### What version of ESLint are you using?

8.37.0

### What version of `eslint-plugin-astro` are you using?

0.26.1

### What did you do?

<details>
<summary>Configuration</summary>

```
module.exports = {
  overrides: [
    {
      files: ['*.astro'],
      parser: 'astro-eslint-parser',
      parserOptions: {
        parser: '@typescript-eslint/parser',
        project: true,
        extraFileExtensions: ['.astro'],
      },
      plugins: ['@typescript-eslint'],
      rules: {
        '@typescript-eslint/no-unused-vars': 'error',
      },
    },
  ],
};
```
</details>

```astro
---
interface Props {
  value: string;
}

const { value } = Astro.props;
---

<div>{value}</div>
```

### What did you expect to happen?

Although `Props` _seems_ unused because it's not exported or referenced anywhere in the file, it's not. Astro uses it inside the file under the hood.

### What actually happened?

```plaintext
/Users/josh/repos/joshuakgoldberg-dot-com-next/src/components/Example.astro
  2:11  warning  'Props' is defined but never used  @typescript-eslint/no-unused-vars
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-xv5ffu?file=README.md&file=.eslintrc.cjs&file=src%2Fcomponents%2FExample.astro&on=stackblitz

### Additional comments

I was directed here from the Astro Discord's `#feedback-ideas` channel (thanks Erika!): https://discord.com/channels/830184174198718474/872579324446928896/1094254255117971476

**Comment by ota-meshi** [comment]
Thank you for posting issue.
Hmm. I think the no-unused-vars are not marked as unused because the parser adds a virtual reference to the props.
However, `@typescript-eslint/no-unused-vars` seems to consider same-position references as recursive references and mark them as unused.
(See comments below as to why we're adding a virtual reference.)
https://github.com/typescript-eslint/typescript-eslint/issues/4508#issuecomment-1030508403

We may need to flag eslintUsed at the same time we add the virtual reference 🤔.
